truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/impl/TVMCI.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2012, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.truffle.api.impl;
+
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.RootNode;
+
+/**
+ * An interface between Truffle API and hosting virtual machine. Not interesting for regular Truffle
+ * API users. Acronym for Truffle Virtual Machine Compiler Interface.
+ *
+ * @since 0.12
+ */
+public abstract class TVMCI {
+    /**
+     * Only useful for virtual machine implementors.
+     *
+     * @since 0.12
+     */
+    protected TVMCI() {
+        // export only for com.oracle.graal.truffle and com.oracle.truffle.api.impl
+        assert getClass().getPackage().getName().equals("com.oracle.graal.truffle") || getClass().getPackage().getName().equals("com.oracle.truffle.api.impl");
+    }
+
+    // the class is a member class to not be (easily) visible in Javadoc
```

Comment:
Please merge #148 once verified.

Re. hardcoding: Yes, it may complicate refactorings. On the other hand, if there is a change in packaging, the unit tests are likely to discover that quite easily. I like that such check really creates "friend API" - e.g. enforces the list of "friends" that can implement the class.